pkg/compactor/partition_compaction_planner.go

```diff
@@ -2,30 +2,146 @@ package compactor
+	// claimed same partition in grouper at same time.
+	time.Sleep(p.plannerDelay)
+
+	visitMarker := newPartitionVisitMarker(p.ringLifecyclerID, partitionedGroupID, partitionID)
+	visitMarkerManager := NewVisitMarkerManager(p.bkt, p.logger, p.ringLifecyclerID, visitMarker)
+	existingPartitionVisitMarker := &partitionVisitMarker{}
+	err := visitMarkerManager.ReadVisitMarker(p.ctx, existingPartitionVisitMarker)
+	visitMarkerExists := true
+	if err != nil {
+		if errors.Is(err, errorVisitMarkerNotFound) {
+			visitMarkerExists = false
+		} else {
+			return nil, fmt.Errorf("unable to get visit marker file for partition with partition ID %d, partitioned group ID %d: %s", partitionID, partitionedGroupID, err.Error())
+		}
+	}
+	if visitMarkerExists {
+		if existingPartitionVisitMarker.GetStatus() == Completed {
+			p.compactorMetrics.compactionsNotPlanned.WithLabelValues(p.userID, cortexMetaExtensions.TimeRangeStr()).Inc()
+			level.Warn(p.logger).Log("msg", "partition is in completed status", "partitioned_group_id", partitionedGroupID, "partition_id", partitionID, "compactor_id", p.ringLifecyclerID, existingPartitionVisitMarker.String())
+			return nil, plannerCompletedPartitionError
+		}
+		if !existingPartitionVisitMarker.IsPendingByCompactor(p.partitionVisitMarkerTimeout, partitionID, p.ringLifecyclerID) {
+			p.compactorMetrics.compactionsNotPlanned.WithLabelValues(p.userID, cortexMetaExtensions.TimeRangeStr()).Inc()
+			level.Warn(p.logger).Log("msg", "partition is not visited by current compactor", "partitioned_group_id", partitionedGroupID, "partition_id", partitionID, "compactor_id", p.ringLifecyclerID, existingPartitionVisitMarker.String())
+			return nil, plannerVisitedPartitionError
+		}
+	}
+
+	// Ensure all blocks fits within the largest range. This is a double check
+	// to ensure there's no bug in the previous blocks grouping, given this Plan()
+	// is just a pass-through.
+	// Modified from https://github.com/cortexproject/cortex/pull/2616/files#diff-e3051fc530c48bb276ba958dd8fadc684e546bd7964e6bc75cef9a86ef8df344R28-R63
+	largestRange := p.ranges[len(p.ranges)-1]
+	rangeStart := getRangeStart(metasByMinTime[0], largestRange)
+	rangeEnd := rangeStart + largestRange
+	noCompactMarked := p.noCompBlocksFunc()
+	resultMetas := make([]*metadata.Meta, 0, len(metasByMinTime))
+
+	for _, b := range metasByMinTime {
+		if b.ULID == DUMMY_BLOCK_ID {
+			continue
+		}
+		blockID := b.ULID.String()
+		if _, excluded := noCompactMarked[b.ULID]; excluded {
+			continue
+		}
+
+		if b.MinTime < rangeStart || b.MaxTime > rangeEnd {
+			return nil, fmt.Errorf("block %s with time range %d:%d is outside the largest expected range %d:%d", blockID, b.MinTime, b.MaxTime, rangeStart, rangeEnd)
+		}
+
+		resultMetas = append(resultMetas, b)
+	}
+
+	if len(resultMetas) < 1 {
+		level.Info(p.logger).Log("msg", "result meta size is empty", "partitioned_group_id", partitionedGroupID, "partition_id", partitionID, "size", len(resultMetas))
```

Comment:
Should this also count as `compactionsNotPlanned`? Is there a case this plan will always be grouped and tried to compacted (eg: all blocks has no compaction marker)?

Reply:
You mean we should return error here? We cannot return error because a plan might have all blocks marked as no compact. We should skip this case and return nil error.

Reply:
Disucssed offline. The metric is for error cases not expected to happen often.